[Seeking early feedback] Add mux protocol parser

ddelnano opened the pull request:
This is an early pull request to add mux protocol support (from [Twitter's finagle](https://github.com/twitter/finagle/blob/b593f4065381ddd12e6256c0cbbdd8d9086d42a0/finagle-mux/src/main/scala/com/twitter/finagle/mux/package.scala)). I intend to add support for the [thrift protocol](https://github.com/apache/thrift/blob/master/doc/specs/thrift-binary-protocol.md), but since my c++ skills are weak I'm looking for early feedback on this easier implementation.

## Testing done
- All of the tests (except for unimplemented stitcher ones) are passing
```
vagrant@ubuntu2004:/vagrant$  /opt/pixielabs/bin/bazel test  src/stirling/source_connectors/socket_tracer/protocols/mux:all
INFO: Analyzed 7 targets (0 packages loaded, 0 targets configured).
INFO: Found 4 targets and 3 test targets...
FAIL: //src/stirling/source_connectors/socket_tracer/protocols/mux:stitcher_test (see /home/vagrant/.cache/bazel/_bazel_vagrant/6ad5fdbcbf2eaa93bd62f92333a2e6e5/execroot/px/bazel-out/k8-fastbuild/testlogs/src/stirling/source_connectors/socket_tracer/protocols/mux/stitcher_test/test.log)
INFO: Elapsed time: 0.494s, Critical Path: 0.23s
INFO: 2 processes: 1 internal, 1 linux-sandbox.
INFO: Build completed, 1 test FAILED, 2 total actions
//src/stirling/source_connectors/socket_tracer/protocols/mux:parse_test (cached) PASSED in 0.1s
//src/stirling/source_connectors/socket_tracer/protocols/mux:types_test (cached) PASSED in 0.1s
//src/stirling/source_connectors/socket_tracer/protocols/mux:stitcher_test FAILED in 0.1s
  /home/vagrant/.cache/bazel/_bazel_vagrant/6ad5fdbcbf2eaa93bd62f92333a2e6e5/execroot/px/bazel-out/k8-fastbuild/testlogs/src/stirling/source_connectors/socket_tracer/protocols/mux/stitcher_test/test.log

INFO: Build completed, 1 test FAILED, 2 total actions

```

## Todo
- [ ] Integrate mux parser with rest of codebase and determine how to end to end test
- [ ] Finish off stitch implementation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pixie-io/pixie?pullRequest=322) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pixie-io/pixie?pullRequest=322) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pixie-io/pixie?pullRequest=322) it.</sub>

oazizi000 reviewed (commented):
@ddelnano Thanks again for this great contribution!!

I left a few scattered comments on some of the files (haven't gone through evertthing yet) just to give you some preliminary feedback.

In terms of getting this contributed into Stirling, we can stage it out into a separate when you're ready:
1) The frame parser
2) The stitcher
3) The integration (which we can help with)

ddelnano commented on src/stirling/source_connectors/socket_tracer/protocols/mux/parse.cc:
```diff
@@ -0,0 +1,112 @@
+  PL_ASSIGN_OR(expr, val_or, return ParseState::kNeedsMoreData)
+
+ParseState ParseFullFrame(BinaryDecoder* decoder, message_type_t type, std::string_view* buf, Frame* frame) {
+  PL_UNUSED(type);
```
Fixed.

ddelnano commented on src/stirling/source_connectors/socket_tracer/protocols/mux/parse.cc:
```diff
@@ -0,0 +1,112 @@
+namespace mux {
+
+#define PL_ASSIGN_OR_RETURN_NEEDS_MORE_DATA(expr, val_or) \
+  PL_ASSIGN_OR(expr, val_or, return ParseState::kNeedsMoreData)
```
Yea, I thought I was using this incorrectly, but that makes sense. Updated.

ddelnano commented on src/stirling/source_connectors/socket_tracer/protocols/mux/parse.cc:
```diff
@@ -0,0 +1,112 @@
+
+  PL_ASSIGN_OR_RETURN_NEEDS_MORE_DATA(uint16_t tagFirst, decoder->ExtractInt<uint16_t>());
+  PL_ASSIGN_OR_RETURN_NEEDS_MORE_DATA(uint8_t tagLast, decoder->ExtractInt<uint8_t>());
+  frame->tag = (tagFirst << 8) | tagLast;
```
Yep, that's what I was doing here. Added a TODO comment.

ddelnano commented on src/stirling/source_connectors/socket_tracer/protocols/mux/parse.cc:
```diff
@@ -0,0 +1,112 @@
+  }
+  // TODO: Add support for reading the Rdispatch reply status
+
+  PL_ASSIGN_OR_RETURN_NEEDS_MORE_DATA(int16_t numCtx, decoder->ExtractInt<int16_t>());
```
Updated.

ddelnano commented on src/stirling/source_connectors/socket_tracer/protocols/mux/parse.cc:
```diff
@@ -0,0 +1,112 @@
+    context.insert({std::string(ctxKey), unpackedValue});
+  }
+
+  frame->context = context;
```
I had seen that used in places, but reading the `std::move` docs didn't make me sure I knew when it was appropriate. Updated to use `std::move`.

ddelnano commented on src/stirling/source_connectors/socket_tracer/protocols/mux/types.h:
```diff
@@ -0,0 +1,109 @@
+  return mux_type.has_value();
+}
+
+/**
```
Updated the comment string with the endian-ness

ddelnano commented on src/stirling/source_connectors/socket_tracer/protocols/mux/types.h:
```diff
@@ -0,0 +1,109 @@
+namespace protocols {
+namespace mux {
+
+enum class Type : int8_t {
```
This is the terminology that finagle's mux implementation uses so I'm not sure what alternatives to use here. If I added a comment that explains this enum in more detail would that suffice?

ddelnano replied:
More details on the mux terminology [here](https://github.com/twitter/finagle/blob/81dbd7966148f080d87c50271316ef1d7891618e/finagle-mux/src/main/scala/com/twitter/finagle/mux/package.scala#L43).

ddelnano commented on src/stirling/source_connectors/socket_tracer/protocols/mux/parse.cc:
```diff
@@ -0,0 +1,112 @@
+  frame->tag = (tagFirst << 8) | tagLast;
+
+  if (frame->type == static_cast<int8_t>(Type::RerrOld) || frame->type == static_cast<int8_t>(Type::Rerr)) {
+    frame->why = buf->substr(8, buf->length());
```
I've updated it to use `ExtractString`, however, I don't see the `frame->len` property / method in the the [FrameBase](https://github.com/pixie-io/pixie/blob/6504c17ca3be693c4ac4f2bbb03b6bf1a7aa2912/src/stirling/source_connectors/socket_tracer/protocols/common/event_parser.h#L43). Can you explain where that method / property exists?

ddelnano commented on src/stirling/source_connectors/socket_tracer/protocols/mux/parse.cc:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  if (frame->type == static_cast<int8_t>(Type::Rinit) || frame->type == static_cast<int8_t>(Type::Tinit)) {
+      // TODO: Add support for reading Tinit and Rinit compression, tls and other parameters
```
I'll update this but I need more clarification on the thread we are discussing using `frame->len`.

ddelnano commented on src/stirling/source_connectors/socket_tracer/protocols/mux/parse.cc:
```diff
@@ -0,0 +1,112 @@
+  if (frame->type == static_cast<int8_t>(Type::Rinit) || frame->type == static_cast<int8_t>(Type::Tinit)) {
+      // TODO: Add support for reading Tinit and Rinit compression, tls and other parameters
+  }
+  // TODO: Add support for reading the Rdispatch reply status
```
I'll update this but I need more clarification on the thread we are discussing using `frame->len`.

ddelnano commented:
@oazizi000 I've addressed your feedback and left a few questions for the comments I wasn't able to resolve.

I've also completed the stitcher implementation.

ddelnano commented on src/stirling/source_connectors/socket_tracer/protocols/mux/types.h:
```diff
@@ -0,0 +1,124 @@
+    if (req_type == Type::Tdrain) return Type::Rdrain;
+    if (req_type == Type::Tdispatch) return Type::Rdispatch;
+    if (req_type == Type::TdiscardedOld || req_type == Type::Tdiscarded) return Type::Rdiscarded;
+    return Type::Tlease;
```
This shouldn't return `Type::Tlease` by default otherwise new mux message types will be incorrectly consumed. Is there a pattern in this code base to return an optional type? It seems like `StatusOr<Type>` might fit?

If we don't find a match we want to ignore the message.
